pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -203,19 +203,29 @@ func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
 		return
 	}
 
-	podKeys := sets.String{}
-	if oldSvcAnnotation != newSvcAnnotation {
-		// Process Pods corresponding to Service with valid NPL annotation.
-		if oldSvcAnnotation == "true" {
-			podKeys = sets.NewString(c.getPodsFromService(oldSvc)...)
-		} else if newSvcAnnotation == "true" {
+	var podKeys sets.String
+	// Check for annotation changes in the Service. Pod sets don't change
+	// in case of annotation and Service type updates.
+	if oldSvcAnnotation != newSvcAnnotation &&
+		(oldSvcAnnotation == "true" || newSvcAnnotation == "true") {
+		podKeys = sets.NewString(c.getPodsFromService(newSvc)...)
+	}
+
+	// Check for Service type changes, NodePort to non-NodePort and vice versa.
+	if oldSvc.Spec.Type != newSvc.Spec.Type &&
+		(oldSvc.Spec.Type == corev1.ServiceTypeNodePort || newSvc.Spec.Type == corev1.ServiceTypeNodePort) {
+		if podKeys == nil {
 			podKeys = sets.NewString(c.getPodsFromService(newSvc)...)
 		}
-	} else if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
+	}
+
+	if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
 		// Disjunctive union of Pods from both Service sets.
 		oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
-		newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
-		podKeys = utilsets.SymmetricDifference(oldPodSet, newPodSet)
+		if podKeys == nil {
+			podKeys = sets.NewString(c.getPodsFromService(newSvc)...)
+		}
+		podKeys = utilsets.SymmetricDifference(oldPodSet, podKeys)
```

Comment:
If annotation/Type is updated as well as Selector, don't it need to process the pods in the intersection?
For example: update a service's annotation from true to false and its selector from app=foo to tier=frontend, shouldn't all pods matching app=foo be processed?